### ISSUE TITLE
No "vectorization" in `f.x`?

### ISSUE THREAD

**Issue by XiaomoWu** [h2oai/datatable#1840]
In `Rdatatable`, columns are treated as vectors, so applying a function equals to applying to every element in the vector:
```
dt = data.table(x=c(1, 2, 3))
dt[, max(x)] # return 3
```
However, in `pydatatable`, if we apply `np.max` to the same column, we get one result for every row, which means the column is not treated as a vector!
```
df = dt.Frame(x=[1, 2, 3])
df[:, np.max(x)] # return [1, 2, 3]
```
Of course this will not be an issue since `pydatatable` has implemented `dt.max`. However, does that mean we can't use any vectorized function (`numpy` has a lot) unless `pydatatable` implements its own version?

**Edit**
A related question is how the `pydatatable` functions compare with their `numpy` counterparts in performance, since the `numpy` functions are highly optimized.

**Comment by st-pasha** [comment]
In R data.table, the expression `dt[, max(x)]` is evaluated like this: 
- first, R calls `data.table.[` operator on `dt`, passing it `(, max(x))` as the argument. 
- then data.table expands that argument into an AST, then traverses that AST replacing `x` with the column "x" from `dt`;
- then that column is passed to R's builtin `max()`, which produces a 1-element vector;
- finally, the result of `max()` is turned back into a `data.table` object.

In Python, there are no ASTs and no lazy evaluation. So when you write `df[:, np.max(f.x)]`, here's what happens: 
- first `f.x` is evaluated, creating a `ColSelectorExpr(f.x)` object. This is a special `datatable` object that is used to mimic lazy evaluation;
- then, this object is passed to numpy function `np.max(...)`. The numpy function, bless its heart, just returns the argument itself. Thus, `np.max(f.x)` becomes simply `f.x`;
- then, tuple `(:, f.x)` is passed to `datatable.Frame.__getitem__`, which interprets it as selection of column "x".

In short, neither builtin `max` nor `np.max` will work for `datatable`, which is why we had to implement our own version `dt.max`.

**Comment by XiaomoWu** [comment]
Thank you for the clear explanation @st-pasha! Does that mean we can't apply custom vectorized function to columns unless there's a `pydatatable` version? For example:
```Python
df = dt.Frame(x=[1, 2, 3])

def add_one(x: list):
    return [_+1 for _ in x]

df[:, add_one(f.x)] # fail
```

**Comment by st-pasha** [comment]
This functionality can be added, but we have not done so thus far.
The major problem here would be the performance: for a custom function the value will have to be wrapped into a python object, passed to the function, then the resulting PyObject would have to be converted back to C++ primitives. All of this would only be able to run in single-threaded mode.
In short, it is no different than doing simply
```
dt.Frame([add_one(v) for v in df[:, f.x].to_list()[0]])
```
Though, if performance doesn't really matter, then having a nicer syntactic sugar is convenient.

**Comment by Victsz** [comment]
Is this functionality or syntactic sugar on the roadmap?

**Comment by st-pasha** [comment]
@Victsz see issue #1960
